airflow/models/dagrun.py

```diff
@@ -804,18 +806,42 @@ def verify_integrity(self, session: Session = NEW_SESSION):
                 ti.state = State.NONE
             session.merge(ti)
 
-        # check for missing tasks
-        for task in dag.task_dict.values():
-            if task.start_date > self.execution_date and not self.is_backfill:
-                continue
+        def task_filter(task: "BaseOperator"):
+            return task.task_id not in task_ids and (
+                self.is_backfill or task.start_date <= self.execution_date
+            )
+
+        created_counts: Dict[str, int] = Counter()
```

Comment:
Since we are just manually incrementing the counts anyway, it would be better to use a `defaultdict(int)` here instead.

Reply:
But `Counter` would be more useful if you change the below to something like

```python
def create_ti(task: "BaseOperator") -> TI:
    ti = TI(task, run_id=self.run_id)
    task_instance_mutation_hook(ti)
    return ti

if hook_is_noop:
    session.bulk_insert_mappings(TI, (TI.insert_mapping(self.run_id, t) for t in tasks))
else:
    session.bulk_save_objects(create_ti(t) for t in tasks)
created_counts.update(t.task_type for t in tasks)
```

Reply:
Yeah, I had it this way before (hence why Counter in the first place) but doing it this way seemed a lot slower. Let me test that again.

Reply:
If Counter is slower (or comparable) I’d just use defaultdict. Counter is cool, but generally does not produce better code, from my experience, and this is no exception.

Reply:
I think the slow down wasn't from using Counter, but just having to walk the list of 20k items twice. Anyway, defaultdict it is.